### ISSUE TITLE
[seq] can't reach end of track

### ISSUE THREAD

**Issue by porres** [porres/pd-cyclone#557]
cyclone/seq ignores 'End of Track' event and does output a bang before the track actually ends. This ignores rests in the end of the MIDI track and is really bad for looping. According to the MIDI specification:
![eot](https://user-images.githubusercontent.com/5360142/132377748-24669334-cd95-4e07-972c-973c47f76ca1.png)

'End of Track' is defined here on our code, but not used effectively https://github.com/porres/pd-cyclone/blob/master/shared/control/mifi.h#L35
 

see also previous discussion at https://github.com/porres/pd-else/issues/1070

**Comment by porres** [comment]
here's a test MIDI file

[one-bar.mid.zip](https://github.com/porres/pd-cyclone/files/7122972/one-bar.mid.zip)

and its score
<img width="903" alt="score" src="https://user-images.githubusercontent.com/5360142/132378198-c1fbba3c-eff4-46af-a48c-9cb33b771064.png">

**Comment by styk-tv** [comment]
one-bar.mid Iincluded above
py-midicsv output, in this version track is ending with the end of the 4th 8th note
```
0, 0, Header, 0, 1, 96
1, 0, Start_track
1, 0, Title_t, "\000"
1, 0, Time_signature, 4, 2, 36, 8
1, 0, Time_signature, 4, 2, 36, 8
1, 0, Note_on_c, 0, 60, 100
1, 48, Note_off_c, 0, 60, 64
1, 96, Note_on_c, 0, 67, 100
1, 144, Note_off_c, 0, 67, 64
1, 192, Note_on_c, 0, 64, 100
1, 240, Note_off_c, 0, 64, 64
1, 288, Note_on_c, 0, 62, 100
1, 336, Note_off_c, 0, 62, 64
1, 336, End_track
0, 0, End_of_fil
```

[one-bar-4x-8th-notes.384.mid.zip](https://github.com/porres/pd-cyclone/files/7140010/one-bar-4x-8th-notes.384.mid.zip)
 in this version track is ending with the end of bar. calculated as follows: 
(track=1 bar=4xPPQ=4x96=384)

py-midicsv output (notes.384),
```
0, 0, Header, 0, 1, 96
1, 0, Start_track
1, 0, Title_t, "\000"
1, 0, Time_signature, 4, 2, 36, 8
1, 0, Time_signature, 4, 2, 36, 8
1, 0, Note_on_c, 0, 60, 100
1, 48, Note_off_c, 0, 60, 64
1, 96, Note_on_c, 0, 67, 100
1, 144, Note_off_c, 0, 67, 64
1, 192, Note_on_c, 0, 64, 100
1, 240, Note_off_c, 0, 64, 64
1, 288, Note_on_c, 0, 62, 100
1, 336, Note_off_c, 0, 62, 64
1, 384, End_track
0, 0, End_of_file
```
Both file play the same, cut the bar short and skip last 8th pause and ending immediately with the end of 4th 8th note.
Conclusion EndOfTrack value is not respected. Please correct me if I'm wrong.

Note from Midi Spec:
![image](https://user-images.githubusercontent.com/838921/132770277-676171ec-4ae7-467a-b516-0768edaa986c.png)

**Comment by porres** [comment]
When opening the MIDI file, first [seq] analyzes the file before actually opening, this is done here https://github.com/porres/pd-cyclone/blob/cd057b7608b5087a2fed651ed9a609f54f2ddb98/shared/control/mifi.c#L862

When analyzing, [seq] checks all MIDI events and counts the number of events here https://github.com/porres/pd-cyclone/blob/cd057b7608b5087a2fed651ed9a609f54f2ddb98/shared/control/mifi.c#L911

But the End of Track is not accounted there as an event.

Later on the playhead follows the number of events and skips this last 'rest' event

**Comment by porres** [comment]
@styk-tv and @MikeMorenoDSP, please test this fix => https://github.com/porres/pd-cyclone/issues/557 so I can close this

**Comment by porres** [comment]
new commit that deals with an error => https://github.com/porres/pd-cyclone/commit/b401ffd4de372b3e38b181b1f818d8948aa57f8f

**Comment by porres** [comment]
Guess what? MAX can't handle 'end of track' either, hahahahahaha...

well, I'm pretty confident I've fixed this, plus I made other improvements, the same are being made to else/midi and ELSE will probably be updated sooner

**Comment by styk-tv** [comment]
thank you @porres

would you be able to trigger a release candidate RC build please?

**Comment by porres** [comment]
If you can't compile, there'll be a new release once pd 0.52 is out, and it was supposed to come out this summer. ELSE must be updated earlier perhaps, but not later ;)

**Comment by styk-tv** [comment]
@porres i can confirm EndOfTrack works as expected.  thank you for your time on this.

I can confirm all I had to do was run `make` in pd-cyclone folder, it produced all the pd_darwin files. To confirm I installed pd-cyclones from official repo in pd, added above one-bar-4x-8th-notes.384.mid example, then removed seq.pd_fat file to make sure it stops working all together, then added newly compiled seq.pd_darwin and bug dissapeared.

Described issue in the ticket is solved.  Attaching built mac binary into this ticket.

[seq.pd_darwin.zip](https://github.com/porres/pd-cyclone/files/7148797/seq.pd_darwin.zip)
